sdks/go/test/integration/io/xlang/bigquery/bigquery.go

```diff
@@ -0,0 +1,114 @@
+	// Number of random elements to create for test. Must be less than number of words in text.
+	inputSize = 50
+)
+
+// TestRow is a sample row to write and read from that is expected to contain enough deterministic
+// and random data in different data types to provide a reasonable signal that reading and writing
+// works at a basic level.
+type TestRow struct {
+	Counter   int64    `beam:"counter"`   // A deterministic counter, increments for each row generated.
+	Rand_data RandData `beam:"rand_data"` // An inner struct containing randomized data.
+}
+
+func shuffleText() []string {
+	words := strings.Fields(text)
+	rand.Shuffle(len(words), func(i, j int) { words[i], words[j] = words[j], words[i] })
+	return words
+}
+
+// RandData is a struct of various types of random data.
+type RandData struct {
+	Flip bool   `beam:"flip"` // Flip is a bool with a random chance of either result (a coin flip).
+	Num  int64  `beam:"num"`  // Num is a random int64.
+	Word string `beam:"word"` // Word is a randomly selected word from a sample text.
+}
+
+// ddlSchema is a string for BigQuery data definition language that corresponds to TestRow.
+const ddlTestRowSchema = "counter INT64 NOT NULL, " +
+	"rand_data STRUCT<" +
+	"flip BOOL NOT NULL," +
+	"num INT64 NOT NULL," +
+	"word STRING NOT NULL" +
+	"> NOT NULL"
+
+// createTestRows creates a number of TestRows, populating the randomized data.
+func createTestRows(_ []byte, emit func(TestRow)) {
+	rand.Seed(time.Now().UnixNano())
+	words := shuffleText()
+	for i := 0; i < inputSize; i++ {
+		emit(TestRow{
+			Counter: int64(i),
+			Rand_data: RandData{
+				Flip: rand.Int63n(2) != 0,
+				Num:  rand.Int63(),
+				Word: words[i],
+			},
+		})
+	}
+}
+
+func WritePipeline(expansionAddr, table string, createFn interface{}) *beam.Pipeline {
+	p := beam.NewPipeline()
+	s := p.Root()
+
+	// Generate elements and write to table.
+	rows := beam.ParDo(s, createFn, beam.Impulse(s))
+	bigqueryio.Write(s, table, rows,
+		bigqueryio.CreateDisposition(bigqueryio.CreateNever),
+		bigqueryio.WriteExpansionAddr(expansionAddr))
+
+	return p
+}
+
+func ReadPipeline(expansionAddr, table string, createFn interface{}) *beam.Pipeline {
+	p := beam.NewPipeline()
+	s := p.Root()
+
+	// Read from table and compare to generated elements.
+	rows := beam.ParDo(s, createFn, beam.Impulse(s))
+	inType := reflect.TypeOf((*TestRow)(nil)).Elem()
+	readRows := bigqueryio.Read(s, inType,
+		bigqueryio.FromTable(table),
```

Comment:
Also test reading from a query.

Reply:
I did. As an update to what we talked about online, I'm not getting any permissions errors testing locally. (I am getting a different unrelated error, but it's because of a bug in how I wrote the test. Will commit once I fix that.)